File: app/src/main/java/com/krystianwsul/checkme/gui/instances/DayFragment.java
package com.krystianwsul.checkme.gui.instances;


import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.FloatingActionButton;
import android.support.design.widget.TabLayout;
import android.support.v4.app.LoaderManager;
import android.support.v4.content.Loader;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.krystianwsul.checkme.R;
import com.krystianwsul.checkme.gui.AbstractFragment;
import com.krystianwsul.checkme.gui.FabUser;
import com.krystianwsul.checkme.gui.MainActivity;
import com.krystianwsul.checkme.gui.instances.tree.GroupListFragment;
import com.krystianwsul.checkme.loaders.DayLoader;
import com.krystianwsul.checkme.utils.time.Date;

import junit.framework.Assert;

import java.text.DateFormatSymbols;
import java.util.Calendar;

public class DayFragment extends AbstractFragment implements LoaderManager.LoaderCallbacks<DayLoader.Data>, FabUser {
    private static final String POSITION_KEY = "position";
    private static final String TIME_RANGE_KEY = "timeRange";

    private int mPosition;
    private MainActivity.TimeRange mTimeRange;

    private TabLayout mDayTabLayout;
    private GroupListFragment mGroupListFragment;

    @Nullable
    private FloatingActionButton mFloatingActionButton;

    @NonNull
    public static DayFragment newInstance(@NonNull MainActivity.TimeRange timeRange, int day) {
        Assert.assertTrue(day >= 0);

        DayFragment dayFragment = new DayFragment();

        Bundle args = new Bundle();
        args.putInt(POSITION_KEY, day);
        args.putSerializable(TIME_RANGE_KEY, timeRange);

        dayFragment.setArguments(args);
        return dayFragment;
    }

    public DayFragment() {

    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_day, container, false);
        Assert.assertTrue(view != null);

        mDayTabLayout = (TabLayout) view.findViewById(R.id.day_tab_layout);
        Assert.assertTrue(mDayTabLayout != null);

        return view;
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);

        Bundle args = getArguments();
        Assert.assertTrue(args != null);

        Assert.assertTrue(args.containsKey(POSITION_KEY));
        mPosition = args.getInt(POSITION_KEY);
        Assert.assertTrue(mPosition >= 0);

        Assert.assertTrue(args.containsKey(TIME_RANGE_KEY));
        mTimeRange = (MainActivity.TimeRange) args.getSerializable(TIME_RANGE_KEY);
        Assert.assertTrue(mTimeRange != null);

        String title;
        if (mTimeRange == MainActivity.TimeRange.DAY) {
            switch (mPosition) {
                case 0:
                    title = getActivity().getString(R.string.today);
                    break;
                case 1:
                    title = getActivity().getString(R.string.tomorrow);
                    break;
                default:
                    Calendar calendar = Calendar.getInstance();

                    calendar.add(Calendar.DATE, mPosition);
                    Date date = new Date(calendar);
                    title = date.getDayOfWeek().toString() + ", " + date.toString();
            }
        } else {
            if (mTimeRange == MainActivity.TimeRange.WEEK) {
                Calendar start = Calendar.getInstance();

                if (mPosition > 0) {
                    start.add(Calendar.WEEK_OF_YEAR, mPosition);
                    start.set(Calendar.DAY_OF_WEEK, start.getFirstDayOfWeek());
                }

                Calendar end = Calendar.getInstance();

                end.add(Calendar.WEEK_OF_YEAR, mPosition + 1);
                end.set(Calendar.DAY_OF_WEEK, end.getFirstDayOfWeek());
                end.add(Calendar.DATE, -1);

                Date startDate = new Date(start);
                Date endDate = new Date(end);

                title = startDate.toString() + " - " + endDate.toString();
            } else {
                Assert.assertTrue(mTimeRange == MainActivity.TimeRange.MONTH);

                Calendar calendar = Calendar.getInstance();
                calendar.add(Calendar.MONTH, mPosition);

                int month = calendar.get(Calendar.MONTH);

                title = DateFormatSymbols.getInstance().getMonths()[month];
            }
        }

        mDayTabLayout.addTab(mDayTabLayout.newTab().setText(title));

        mGroupListFragment = (GroupListFragment) getChildFragmentManager().findFragmentById(R.id.day_frame);
        Assert.assertTrue(mGroupListFragment != null);

        if (mFloatingActionButton != null)
            mGroupListFragment.setFab(mFloatingActionButton);

        getLoaderManager().initLoader(0, null, this);
    }

    @Override
    public Loader<DayLoader.Data> onCreateLoader(int id, Bundle args) {
        return new DayLoader(getActivity(), mPosition, mTimeRange);
    }

    @Override
    public void onLoadFinished(Loader<DayLoader.Data> loader, DayLoader.Data data) {
        Assert.assertTrue(data != null);

        mGroupListFragment.setAll(mTimeRange, mPosition, data.DataId, data.mDataWrapper);
    }

    @Override
    public void onLoaderReset(Loader<DayLoader.Data> loader) {

    }

    public void selectAll() {
        mGroupListFragment.selectAll();
    }

    @Override
    public void setFab(@NonNull FloatingActionButton floatingActionButton) {
        if (mFloatingActionButton == floatingActionButton)
            return;

        mFloatingActionButton = floatingActionButton;

        if (mGroupListFragment != null)
            mGroupListFragment.setFab(floatingActionButton);
    }

    @Override
    public void clearFab() {
        mFloatingActionButton = null;

        mGroupListFragment.clearFab();
    }
}

File: app/src/main/java/com/krystianwsul/checkme/gui/instances/DayFragment.kt
package com.krystianwsul.checkme.gui.instances


import android.content.Context
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.annotation.StringRes
import androidx.appcompat.widget.LinearLayoutCompat
import com.jakewharton.rxrelay3.BehaviorRelay
import com.krystianwsul.checkme.MyApplication
import com.krystianwsul.checkme.Preferences
import com.krystianwsul.checkme.R
import com.krystianwsul.checkme.databinding.FragmentDayBinding
import com.krystianwsul.checkme.gui.main.MainActivity
import com.krystianwsul.checkme.gui.utils.BottomFabMenuDelegate
import com.krystianwsul.checkme.utils.partition
import com.krystianwsul.checkme.utils.time.toDateTimeTz
import com.krystianwsul.checkme.viewmodels.DayViewModel
import com.krystianwsul.common.time.Date
import io.reactivex.rxjava3.core.Maybe
import io.reactivex.rxjava3.core.Observable
import io.reactivex.rxjava3.disposables.CompositeDisposable
import io.reactivex.rxjava3.kotlin.Observables
import io.reactivex.rxjava3.kotlin.addTo
import io.reactivex.rxjava3.kotlin.plusAssign
import java.util.*


class DayFragment @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : LinearLayoutCompat(context, attrs, defStyleAttr) {
    // todo consider subclass

    companion object {

        fun getTitle(position: Int): String {
            fun getString(@StringRes stringId: Int) = MyApplication.instance.getString(stringId)

            return when (position) {
                0 -> getString(R.string.today)
                1 -> getString(R.string.tomorrow)
                else -> {
                    Date(
                        Calendar.getInstance()
                            .apply { add(Calendar.DATE, position) }
                            .toDateTimeTz()
                    ).let { it.dayOfWeek.toString() + ", " + it.toString() }
                }
            }
        }
    }

    private val key = BehaviorRelay.create<Int>()

    private var fabDelegate: BottomFabMenuDelegate.FabDelegate? = null

    private val activity = context as MainActivity

    private val dayViewModel = activity.dayViewModel
    private var entry: DayViewModel.Entry? = null

    private val attachedToWindowDisposable = CompositeDisposable()

    private val compositeDisposable = CompositeDisposable()

    private val binding: FragmentDayBinding

    init {
        check(context is Host)

        layoutParams = ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT)
        orientation = VERTICAL

        binding = FragmentDayBinding.inflate(LayoutInflater.from(context), this)

        binding.groupListFragment.apply {
            listener = (context as MainActivity).daysGroupListListener
            forceSaveStateListener = { saveState() }
        }
    }

    fun saveState() = activity.setState(key.value!!, binding.groupListFragment.onSaveInstanceState())

    fun setPosition(position: Int) {
        entry?.stop()
        compositeDisposable.clear()

        key.value?.let { saveState() }

        key.accept(position)

        activity.getState(key.value!!)?.let {
            binding.groupListFragment.onRestoreInstanceState(it)
        }

        // this seems redundant/obsolete, but I'll leave it for now
        fabDelegate?.let(binding.groupListFragment::setVisible)

        entry = dayViewModel.getEntry(position)
    }

    override fun onAttachedToWindow() {
        super.onAttachedToWindow()

        val hostEvents = key.switchMap { key -> (context as Host).hostEvents.map { Pair(key, it) } }

        val (startEvents, stopEvents) =
            hostEvents.partition { (key, event) -> event is Event.PageVisible && event.position == key }

        compositeDisposable += startEvents.subscribe { (_, event) ->
            check(event is Event.PageVisible)

            setFab(event.fabDelegate)

            activity.selectAllRelay
                .subscribe {
                    binding.groupListFragment
                        .treeViewAdapter
                        .selectAll()
                }
                .addTo(compositeDisposable)
        }

        Observables.combineLatest(startEvents, Preferences.showAssignedObservable)
            .subscribe { (_, showAssigned) -> entry!!.start(showAssigned) }
            .addTo(compositeDisposable)

        compositeDisposable += stopEvents.subscribe {
            entry!!.stop()

            clearFab()
            saveState()
        }

        key.switchMap { key -> entry!!.data.map { Triple(key, it, entry!!.dataId) } }
            .subscribe { (key, data, dataId) ->
                binding.groupListFragment.setAll(
                    key,
                    dataId,
                    data.immediate,
                    data.groupListDataWrapper,
                )
            }
            .addTo(compositeDisposable)

        hostEvents.switchMap { (key, event) ->
            if (event is Event.PageVisible && event.position == key)
                activity.started
            else
                Observable.never()
        }
            .filter { it }
            .subscribe { binding.groupListFragment.checkCreatedTaskKey() }
            .addTo(compositeDisposable)

        // this should be inside the GroupListFragment
        activity.dateChangeRelay
            .switchMapMaybe {
                binding.groupListFragment
                    .treeViewAdapterNullable
                    ?.listUpdates
                    ?.firstElement()
                    ?: Maybe.empty()
            }
            .subscribe { binding.groupListFragment.scrollToTop() }
            .addTo(attachedToWindowDisposable)
    }

    override fun onDetachedFromWindow() {
        compositeDisposable.clear()
        attachedToWindowDisposable.clear()

        super.onDetachedFromWindow()
    }

    private fun setFab(fabDelegate: BottomFabMenuDelegate.FabDelegate) {
        this.fabDelegate = fabDelegate

        binding.groupListFragment.setVisible(fabDelegate)
    }

    private fun clearFab() {
        fabDelegate = null

        binding.groupListFragment.clearFab()
    }

    interface Host {

        val hostEvents: Observable<Event>
    }

    sealed class Event {

        data class PageVisible(val position: Int, val fabDelegate: BottomFabMenuDelegate.FabDelegate) : Event()

        object Invisible : Event()
    }
}
